account_edi_ubl_cii_purchase_match/models/account_edi_xml_ubl_20.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2025 ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import models
+
+
+class AccountEdiXmlUBL20(models.AbstractModel):
+
+    _inherit = "account.edi.xml.ubl_20"
+
+    def _import_fill_invoice_form(self, journal, tree, invoice, qty_factor):
+        invoice = invoice.with_context(no_purchase_set=True)
+        res = super()._import_fill_invoice_form(journal, tree, invoice, qty_factor)
+        if journal.type != "purchase":
+            return res
+        invoice_origin_node = tree.find("./{*}OrderReference/{*}ID")
+        if invoice_origin_node is None:
+            return res
+        order_ref = invoice_origin_node.text
+        self._match_invoice_to_purchase_order(invoice, order_ref)
+        return res
+
+    def _match_invoice_to_purchase_order(self, invoice, order_ref):
+        purchase_order = self.env["purchase.order"].search(
+            [
+                "|",
+                ("name", "=", order_ref),
+                ("partner_ref", "=", order_ref),
+                ("state", "in", ("purchase", "done")),
+            ],
+            limit=1,
+        )
+        if not purchase_order:
+            return False
+        for invoice_line in invoice.invoice_line_ids:
+            self._match_invoice_line_to_purchase_order_line(
+                invoice_line, purchase_order.order_line
+            )
+        return True
+
+    def _match_invoice_line_to_purchase_order_line(self, invoice_line, purchase_lines):
+        if invoice_line.purchase_line_id:
+            return False
+        product = self._get_matching_product(invoice_line)
+        purchase_line = purchase_lines.filtered(lambda line: line.product_id == product)
```

Comment:
What if no product is found and there are PO lines without product?
```suggestion
        purchase_line = purchase_lines.filtered(lambda line: line.product_id and line.product_id == product)
```